### PR TITLE
python: attempts to fix 3d mat parsing problem for dnn

### PR DESCRIPTION
**Pull request by fengyuentau**
Fixes https://github.com/opencv/opencv/issues/25762 https://github.com/opencv/opencv/issues/23242
Relates https://github.com/opencv/opencv/issues/25763 https://github.com/opencv/opencv/issues/19091

Although `cv.Mat` has already been introduced to workaround this problem, people do not know it and it kind of leads to confusion with `numpy.array`. This patch adds a "switch" to turn off the auto multichannel feature when the API is from cv::dnn::Net (more specifically, `setInput`) and the parameter is of type `Mat`. This patch only leads to changes of three places in `pyopencv_generated_types_content.h`:

```.diff
static PyObject* pyopencv_cv_dnn_dnn_Net_setInput(PyObject* self, PyObject* py_args, PyObject* kw)
{
...
- pyopencv_to_safe(pyobj_blob, blob, ArgInfo("blob", 0)) &&
+ pyopencv_to_safe(pyobj_blob, blob, ArgInfo("blob", 8)) &&
...
}

// I guess we also need to change this as one-channel blob is expected for param
static PyObject* pyopencv_cv_dnn_dnn_Net_setParam(PyObject* self, PyObject* py_args, PyObject* kw)
{
...
- pyopencv_to_safe(pyobj_blob, blob, ArgInfo("blob", 0)) )
+ pyopencv_to_safe(pyobj_blob, blob, ArgInfo("blob", 8)) )
...
- pyopencv_to_safe(pyobj_blob, blob, ArgInfo("blob", 0)) )
+ pyopencv_to_safe(pyobj_blob, blob, ArgInfo("blob", 8)) )
...
}
```

Others are unchanged, e.g. `dnn_SegmentationModel` and stuff like that.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/python/src2/gen2.py [hunk 1]
```diff
@@ -598,14 +600,15 @@ def __init__(self, namespace, classname, name, decl, isconstructor, known_classe
             self.rettype = ""
         self.args = []
         self.array_counters = {}
+        self.mat_single_channel = (namespace == "cv.dnn" and classname == "dnn_Net")
```
Comment (asmorkalov) [inline modules/python/src2/gen2.py:603]: I propose to extract it from implementation and put as array like here: https://github.com/opencv/opencv/blob/2028998e00ab94fa8e2e9d3d888cc2d769d25027/modules/python/src2/typing_stubs_generation/predefined_types.py#L20

**Comment by asmorkalov** [comment]
@fengyuentau should work for DNN as the most popular case, but what about other functions, e.g. contrib? There is no way to define the same behaviour there.

File: modules/python/src2/gen2.py (hunk 1 above)
Comment (opencv-alalek) [reply modules/python/src2/gen2.py:603]: use `misc/python` instead.

Comment (fengyuentau) [reply modules/python/src2/gen2.py:603]: Thank you for your reviews.

---

@opencv-alalek Could you be more specific? Do you mean I should create a file (.py) and save the array there then `gen2.py` load it from there?

**Comment by fengyuentau** [comment]
> @fengyuentau should work for DNN as the most popular case, but what about other functions, e.g. contrib? There is no way to define the same behaviour there.

Could you give an example on functions in contrib? I think, if we extract the filter `self.mat_single_channel = (namespace == "cv.dnn" and classname == "dnn_Net")` and put it in any place we want, we can do the same for other functions as well.

**Comment by vpisarev** [comment]
@fengyuentau, thank you! I'd suggest to add another function/method parameter attribute `CV_INP_ND` to bindings generator (hdr_parser.py) so that Python bindings generator could call ndarray=>cv::Mat conversion with proper flags.

**Comment by fengyuentau** [comment]
> @fengyuentau, thank you! I'd suggest to add another function/method parameter attribute `CV_INP_ND` to bindings generator (hdr_parser.py) so that Python bindings generator could call ndarray=>cv::Mat conversion with proper flags.
> 
> e.g.
> 
> ```
> class Net { ...
> CV_WRAP void setInput(CV_INP_ND InputArray tensor);
> };
> ```

Discussed and decided to use `CV_ND` for the macro.

**Comment by fengyuentau** [comment]
Done refactor with `CV_ND`.

**Comment by asmorkalov** [comment]
@fengyuentau Please add/modify python tests to cover the feature.

File: modules/dnn/misc/python/test/test_dnn.py [hunk 2]
```diff
@@ -455,10 +455,6 @@ def test_input_3d(self):
                                     "Verify OPENCV_DNN_TEST_DATA_PATH configuration parameter.")
 
         input = np.load(input_file)
-        # we have to expand the shape of input tensor because Python bindings cut 3D tensors to 2D
-        # it should be fixed in future. see : https://github.com/opencv/opencv/issues/19091
-        # please remove `expand_dims` after that
-        input = np.expand_dims(input, axis=3)
```
Comment (fengyuentau) [inline modules/dnn/misc/python/test/test_dnn.py:461]: `setInput` has been tested in this one. It was previously hacked to make it passed.

---

I also added `CV_ND` macro for `setParam`. Does it need to be tested? This API is not tested any where though.

File: modules/dnn/misc/python/test/test_dnn.py [hunk 3]
```diff
@@ -474,6 +470,31 @@ def test_input_3d(self):
 
             normAssert(self, real_output, gold_output, "", getDefaultThreshold(target))
 
+    def test_set_param_3d(self):
```
Comment (fengyuentau) [inline modules/dnn/misc/python/test/test_dnn.py:473]: @asmorkalov A new test for `setParam` is added anyway. Tested without `CV_ND` and it fails.

**Review by opencv-alalek — COMMENTED** [review]
Output arrays should be handled too.

E.g. `forward(OutputArrayOfArrays`

**Comment by opencv-alalek** [comment]
There is dedicated header which is used to test binding generator features: https://github.com/opencv/opencv/blob/4.x/modules/core/include/opencv2/core/bindings_utils.hpp

**Comment by fengyuentau** [comment]
> Output arrays should be handled too.
> 
> E.g. `forward(OutputArrayOfArrays`

Specifically speaking for numpy, there is no multi-channel numpy array. So there is no need to mark output.

**Comment by opencv-alalek** [comment]
There is ability to pass pre-allocated arrays via such API. So shapes of these arrays should be properly passed to C++ code.

**Comment by fengyuentau** [comment]
> There is ability to pass pre-allocated arrays via such API. So shapes of these arrays should be properly passed to C++ code.

Unfortunately it seems we cannot do this for `forward(OutputArrayOfArrays output, ...)`, since it is already parsed to be vector of Mat as return value (tuple in Python). See beflow for one of the bindings. ArgInfo flag cannot control this behavior since `pyopencv_to` does not take `ArgInfo`. Unlike `setInput` which we can do it inside `pyopencv_to_safe`, `forward` defines the retval ouside of `pyopencv_to_safe`.


```c++
{
    PyObject* pyobj_outputBlobs = NULL;
    vector_Mat outputBlobs; // output has been defined as vector of Mat
    PyObject* pyobj_outputName = NULL;
    String outputName;

    const char* keywords[] = { "outputBlobs", "outputName", NULL };
    if( PyArg_ParseTupleAndKeywords(py_args, kw, "|OO:dnn_Net.forward", (char**)keywords, &pyobj_outputBlobs, &pyobj_outputName) &&
        pyopencv_to_safe(pyobj_outputBlobs, outputBlobs, ArgInfo("outputBlobs", 1)) &&
        pyopencv_to_safe(pyobj_outputName, outputName, ArgInfo("outputName", 0)) )
    {
        ERRWRAP2(_self_->forward(outputBlobs, outputName));
        return pyopencv_from(outputBlobs);
    }


        pyPopulateArgumentConversionErrors();
    }
```

---

In order to support ND parsing, we will have to spend more time modifying the script of binding generator. People barely use Python interface in the following way to trigger this bahavior, so I guess it makes sense to keep it as is for now:

```
real_output = np.empty([2, 5, 3], dtype=np.float32)
net.forward(real_output, "237") # real_output will be a length-2 tuple with two numpy arrays of shape [5, 3]
```

**Review by asmorkalov — APPROVED** [review]
:+1:

**Comment by fengyuentau** [comment]
> There is ability to pass pre-allocated arrays via such API. So shapes of these arrays should be properly passed to C++ code.

@opencv-alalek I updated code according to your comment. This introduce a convention:

- The pre-allocated is a `np.array`, we treat it as a `Mat` and parse with ND flag.
- The pre-allocated is a `tuple` of `np.array`l, we treat it as a vector of `Mat` and the ND flag can still apply if any.

**Review by VadimLevin — APPROVED** [review]

**Review by opencv-alalek — APPROVED** [review]

**Pull request merged**